### PR TITLE
[xlcore][feature] Dxvk settings rework

### PR DESCRIPTION
**Pull request by rankynbass**
Edit: Links updated for building with marzent's suggestions added.

This PR will add the necessary framework to XIVLauncher.Common.Unix to allow XL Core to add several new features around DXVK. This is only the first part of the patch, but it needs to be in place before XL Core can be updated.

* Moves all DXVK-related settings to a new class - DxvkSettings.
* Adds support for DXVK version switching. Currently this includes 1.10.1 - 1.10.3, and 2.0. Default is 1.10.1, and it pulls from the dxvk-async repo just like before.
* Adds support for DXVK_FRAMERATE_LIMIT. Setting the value to 0 sets framerate to unlimited.
* Adds support for custom DXVK Hud strings. This includes verifying that you only used supported options.
* Adds support for MangoHud
* Actually does some checking on the file paths for MangoHud, so that you acutally have to pass it a file for MangoHud Custom.
* Includes sensible fallbacks in the case that these changes are built against the current XL Core. Only the new DXVK hud options and MangoHud will show up, but it will build and run.
* Change the 4th argument in the constructor method in Launcher.cs to have a default value of null; this is just to allow XL Core to build without changes. 

I've been working on this for a while, trying to get it into a place where it would work in a flatpak environment without hanging or crashing. It's finally working now.

You can try it out for yourself by following the instructions here: [XL Core Testing Flatpak](https://github.com/rankynbass/dev.goats.testing.xivlauncher/tree/new-dxvk-settings), or build it by cloning [my fork of XL Core](https://github.com/rankynbass/XIVLauncher.Core/tree/xlcore-fixes-plus-dxvk) and checking out the xlcore-fixes-plus-dxvk branch, and doing a `dotnet publish`

The `master` branch includes all the XL Core updates I'm working on, and the 1.0.2+new-dxvk-settings branch, is there to prove that it can be built against XL Core with no changes other than updating the submodule.

Here's some screenshots of how it looks (with XL Core patches).

Wine Tab:
dxvk-rework-wine-tab
Dxvk Tab:
dxvk-rework-dxvk-tab
Version Switcher:
dxvk-rework-versions
Hud Switcher:
dxvk-rework-hud-options

### CODE REVIEW

File: src/XIVLauncher.Common/Game/Launcher.cs [hunk 1]
```diff
@@ -40,7 +40,7 @@ public class Launcher
 
     private const string FALLBACK_FRONTIER_URL_TEMPLATE = "https://launcher.finalfantasyxiv.com/v620/index.html?rc_lang={0}&time={1}";
 
-    public Launcher(ISteam? steam, IUniqueIdCache uniqueIdCache, ISettings settings, string? frontierUrl)
+    public Launcher(ISteam? steam, IUniqueIdCache uniqueIdCache, ISettings settings, string? frontierUrl = null)
```
Comment (marzent) [inline src/XIVLauncher.Common/Game/Launcher.cs:43]: This seems out of scope for this PR

File: src/XIVLauncher.Common/Game/Launcher.cs [hunk 2]
```diff
@@ -71,7 +71,7 @@ public Launcher(ISteam? steam, IUniqueIdCache uniqueIdCache, ISettings settings,
         this.client = new HttpClient(handler);
     }
 
-    public Launcher(byte[] steamTicket, IUniqueIdCache uniqueIdCache, ISettings settings, string? frontierUrl)
+    public Launcher(byte[] steamTicket, IUniqueIdCache uniqueIdCache, ISettings settings, string? frontierUrl = null)
```
Comment (marzent) [inline src/XIVLauncher.Common/Game/Launcher.cs:74]: Ditto

File: src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        frameRate ??= 0;
+        if (frameRate > 0) this.DxvkVars.Add("DXVK_FRAME_RATE", (frameRate).ToString());
+        this.DxvkVersion = version;
+        string release = this.DxvkVersion switch
+        {
+            Dxvk.DxvkVersion.v1_10_1 => "1.10.1",
+            Dxvk.DxvkVersion.v1_10_2 => "1.10.2",
+            Dxvk.DxvkVersion.v1_10_3 => "1.10.3",
+            Dxvk.DxvkVersion.v2_0 => "2.0",
+            _ => VersionOutOfRange(),
```
Comment (marzent) [inline src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs:39]: This probably should be an `ArgumentOutOfRangeException`, as it is not really supposed to happen and you'd have to change the default DXVK version in the future in like 3 different places

File: src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                DxvkVars.Add("MANGOHUD","1");
+                DxvkVars.Add("MANGOHUD_CONFIG","full");
+                break;
+            // If DxvkHudType is None, or undefined, don't set anything.
+            default:
+                break;
+        }
+    }
+
+    private string VersionOutOfRange() {
```
Comment (marzent) [inline src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs:96]: This is unneeded with the above change

File: src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    public string FolderName { get; private set; }
+
+    public Dictionary<string, string> DxvkVars { get; private set; }
+
+    public Dxvk.DxvkHudType DxvkHud { get; private set; }
+
+    public Dxvk.DxvkVersion DxvkVersion { get; private set; }
+
+    private const string ALLOWED_CHARS = "^[0-9a-zA-Z,=.]+$";
+    private const string ALLOWED_WORDS = "^(?:devinfo|fps|frametimes|submissions|drawcalls|pipelines|descriptors|memory|gpuload|version|api|cs|compiler|samplers|scale=(?:[0-9]){0,2}(?:.(?:[0-9])+)?)$";
```
Comment (marzent) [inline src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs:21]: Is there a reason for scale being < 100 ?
DXVK will allow such values and will sanitize them by itself, in fact the only restriction it enforces is scale >= 0.01:
https://github.com/doitsujin/dxvk/blob/c0d843c5788448ed66d61246c6ad8453c41c87fd/src/dxvk/hud/dxvk_hud.cpp#L13

File: src/XIVLauncher.Common.Unix/Compatibility/Dxvk.cs [hunk 6]
```diff
@@ -49,7 +47,34 @@ public enum DxvkHudType
...
+
+    public enum DxvkVersion
+    {
+        [SettingsDescription("1.10.1 (default)", "The default version of DXVK used with XIVLauncher.Core.")]
+        v1_10_1,
+
+        [SettingsDescription("1.10.2", "Newer version of 1.10 branch of DXVK. Probably works.")]
+        v1_10_2,
+
+        [SettingsDescription("1.10.3", "Newer version of 1.10 branch of DXVK. Probably works.")]
```
Comment (marzent) [inline src/XIVLauncher.Common.Unix/Compatibility/Dxvk.cs:74]: 1.10.3 for sure works with Dalamud + GShade; it's really only 2.0 that is creating trouble

**Review by marzent — CHANGES_REQUESTED** [review]
Thanks for bringing all DXVK settings together in one class, this is a nice change.

Just a few thoughts on even improving DXVK handling further:
Setting `DXVK_LOG_PATH` to the xlcore provided logs folder would be nice.
In the long run it would also be good to set `DXVK_CONFIG_FILE` and `DXVK_STATE_CACHE_PATH` to outside of the wine prefix, as that would make the shader state cache and DXVK configuration survive a prefix swap or deletion

File: src/XIVLauncher.Common.Unix/Compatibility/CompatibilityTools.cs [hunk 7]
```diff
@@ -68,10 +66,28 @@ public CompatibilityTools(WineSettings wineSettings, Dxvk.DxvkHudType hudType, b
                 this.dxvkDirectory.Create();
         }
 
+        if (!wineSettings.Prefix.Exists)
```
[the remark this answers is not in the document]
Comment (rankynbass) [reply src/XIVLauncher.Common.Unix/Compatibility/CompatibilityTools.cs]: So it is... deleting that. You know how it goes with copy/paste

File: src/XIVLauncher.Common.Unix/Compatibility/CompatibilityTools.cs [hunk 8]
```diff
@@ -68,10 +66,28 @@ public CompatibilityTools(WineSettings wineSettings, Dxvk.DxvkHudType hudType, b
...
+
+            if (!this.dxvkDirectory.Exists)
+                this.dxvkDirectory.Create();
+        }
+
         if (!wineSettings.Prefix.Exists)
             wineSettings.Prefix.Create();
     }
 
+        public CompatibilityTools(WineSettings wineSettings, Dxvk.DxvkHudType hudType, bool? gamemodeOn, bool? dxvkAsyncOn, DirectoryInfo toolsFolder)
```
[the remark this answers is not in the document]
Comment (rankynbass) [reply src/XIVLauncher.Common.Unix/Compatibility/CompatibilityTools.cs]: That's fine. I was just putting it in so it would still build against unmodified xlcore. Deleting this as  well.

File: src/XIVLauncher.Common/Game/Launcher.cs (hunk 1 above)
Comment (rankynbass) [reply src/XIVLauncher.Common/Game/Launcher.cs]: Fair enough. It was just in so I could build against unmodified xlcore. Can definitely remove.

File: src/XIVLauncher.Common/Game/Launcher.cs (hunk 2 above)
Comment (rankynbass) [reply src/XIVLauncher.Common/Game/Launcher.cs]: As above.

File: src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs (hunk 5 above)
Comment (rankynbass) [reply src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs]: I didn't know if there were any limits, it just seemed a reasonable high number. Scales over 10 are probably not useful unless you're on an 8K+ display, but it's easy enough to change {0,2} to *. So... changed.

File: src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs (hunk 3 above)
Comment (rankynbass) [reply src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs]: Alright, found a couple of examples in the code and mimicked them. It compiles, and, as you said, this shouldn't ever actually happen to a normal user. Changed.

File: src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs (hunk 4 above)
Comment (rankynbass) [reply src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs]: And deleted.

**Comment by rankynbass** [comment]
Added the dxvk log to logs folder, and added a .dxvk folder for state cache. It'll even be possible to pass in the correct storage path, assuming we ever get around to allowing changing that. By default, just assumes ~/.xlcore, though.

**Comment by rankynbass** [comment]
Alright, should be good now as of commit df1a3be

**Comment by rankynbass** [comment]
And I'm not sure if sync/async need different caches, so I'm keeping them separate now. Hopefully that's everything.

**Comment by marzent** [comment]
> And I'm not sure if sync/async need different caches, so I'm keeping them separate now. Hopefully that's everything.

Theoretically they shouldn't, but there is some bugginess with async and DXVK 2.0 where async will miss saving entries sometimes, so this is probably cleaner

File: src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    private const string ALLOWED_WORDS = "^(?:devinfo|fps|frametimes|submissions|drawcalls|pipelines|descriptors|memory|gpuload|version|api|cs|compiler|samplers|scale=(?:[0-9])*(?:.(?:[0-9])+)?)$";
+
+    public DxvkSettings(Dxvk.DxvkHudType hud = Dxvk.DxvkHudType.None, string dxvkHudCustom="",
+        string mangoHudPath="", bool? async = true, int? frameRate = 0,
+        Dxvk.DxvkVersion version = Dxvk.DxvkVersion.v1_10_1, string? corePath = null)
+    {
+        this.DxvkHud = hud;
+        string home = Environment.GetEnvironmentVariable("HOME");
+        corePath ??= Path.Combine(home,".xlcore");
+        string dxvkConfigPath = Path.Combine(corePath,".dxvk");
```
Comment (marzent) [inline src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs:30]: Maybe it is cleaner to pass a DirectoryInfo into the cctor here as well, just in case the base xlcore storage path will change

Comment (rankynbass) [reply src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs]: That's just the default. You can pass in the correct path to the constructor; it just defaults to null and sets the default to ~/.xlcore in that case. XL Core's Program.cs creates a DxvkSettings object, and I've passed it in there. But I could change it to pass in a DirectoryInfo. It's probably more "correct". So now the storage.Root object can be passed in from XL Core, instead of the string storage.Root.FullName.

**Comment by rankynbass** [comment]
Pulled in marzent's commits. I am using the public static functions I declared in the XL.Core UI, so I added those back in.

**Comment by rankynbass** [comment]
Got rid of a useless function that marzent previously removed and I foolishly re-added because I panicked when I saw a compile error about missing function `CheckMangoHudPath()`. Since all the function did was call `File.Exists()`, I just had to update XL.Core to use that instead of the function in question.

That should be it for now. Everything seems to be working when I build [PR #17](https://github.com/goatcorp/XIVLauncher.Core/pull/17), both natively and as a flatpak.

**Comment by rankynbass** [comment]
Should we bump the default version to 1.10.3, and leave 1.10.1 and .2 in there for compatibility/testing purposes?

**Comment by marzent** [comment]
> Should we bump the default version to 1.10.3, and leave 1.10.1 and .2 in there for compatibility/testing purposes?

1.10.3 is perfectly fine on any platform and plays nicer with some nvidia drivers iirc

**Review by marzent — APPROVED** [review]

File: src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+            FolderName = $"dxvk-async-{release}";
+            DxvkVars.Add("DXVK_ASYNC", async ? "1" : "0");
+        }
+        else
+        {
+            DownloadURL = $"https://github.com/doitsujin/dxvk/releases/download/v{release}/dxvk-{release}.tar.gz";
+            FolderName = $"dxvk-{release}";
+        }
+
+        DirectoryInfo dxvkCachePath = new DirectoryInfo(Path.Combine(dxvkConfigPath.FullName, "cache"));
```
Comment (marzent) [inline src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs:65]: Very minor nitpick, this is a bit inconsistent with the use of var everywhere else

File: src/XIVLauncher.Common.Unix/Compatibility/CompatibilityTools.cs [hunk 11]
```diff
@@ -168,21 +163,14 @@ private Process RunInPrefix(ProcessStartInfo psi, string workingDirectory, IDict
...
-
-        if (this.gamemodeOn == true && !ldPreload.Contains("libgamemodeauto.so.0"))
+        if (gamemodeOn && !ldPreload.Contains("libgamemodeauto.so.0"))
         {
             ldPreload = ldPreload.Equals("") ? "libgamemodeauto.so.0" : ldPreload + ":libgamemodeauto.so.0";
         }
 
-        wineEnviromentVariables.Add("DXVK_HUD", dxvkHud);
-        wineEnviromentVariables.Add("DXVK_ASYNC", dxvkAsyncOn);
+        foreach (KeyValuePair<string, string> dxvkVar in DxvkSettings.DxvkVars)
```
Comment (marzent) [inline src/XIVLauncher.Common.Unix/Compatibility/CompatibilityTools.cs:171]: same thing here

Comment (rankynbass) [reply src/XIVLauncher.Common.Unix/Compatibility/CompatibilityTools.cs]: Alright, changed those to "var".

File: src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs (hunk 10 above)
Comment (rankynbass) [reply src/XIVLauncher.Common.Unix/Compatibility/DxvkSettings.cs]: Changed.

**Comment by rankynbass** [comment]
Made obsolete by #1366
